criu/files-reg.c

```diff
@@ -1571,6 +1577,111 @@ static int get_build_id(const int fd, const struct stat *fd_status,
 	return ret;
 }
 
+static inline void calculate_checksum_iterator_init(u64 *iter)
+{
+	if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_EVERY)
+		*iter = 0;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST)
+		*iter = 0;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD)
+		*iter = 0;
+}
+
+static inline void calculate_checksum_iterator_next(u64 *iter)
+{
+	if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_EVERY)
+		*iter += 1;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST)
+		*iter += 1;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD)
+		*iter += opts.file_validation_chksm_parameter;
+}
+
+static inline bool calculate_checksum_iterator_stop_condition(u64 iter)
+{
+	if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_EVERY) {
+		return false;
+	} else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST) {
+		if (iter >= opts.file_validation_chksm_parameter)
+			return true;
+		else
+			return false;
+	} else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD) {
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ * Does the actual work of calculating the CRC32C checksum of "some"
+ * parts of the file. These "some" parts depend on the configuration chosen.
+ * By default, the first 1024 bytes of the file or the entire file, which ever
+ * is smaller, is considered.
+ * Returns true if the checksum could be obtained, else it returns false.
+ */
+static bool calculate_checksum(const int fd, const struct stat *fd_status,
+				u32 *checksum)
+{
+	int bit_count;
+	u32 byte, mask;
+	u64 up_bound, low_bound = 0, i = 0;
+	unsigned char *file_header;
+
+	/* At most, the first 10MB of the file is mapped. */
+	up_bound = min_t(size_t, fd_status->st_size, CHKSM_CHUNK_SIZE);
+	file_header = (unsigned char *) mmap(0, up_bound,
+						PROT_READ, MAP_PRIVATE | MAP_FILE, fd, 0);
+	if (file_header == MAP_FAILED) {
+		pr_warn("Couldn't mmap file with fd %d", fd);
+		return false;
+	}
+
+	*checksum = 0xFFFFFFFF;
+	calculate_checksum_iterator_init(&i);
+
+	while (!calculate_checksum_iterator_stop_condition(i) &&
+			i >= low_bound && i < up_bound && i < fd_status->st_size) {
```

Comment:
This would be easier to understand if it was for () cycle.
You don't need checking low_bound every loop as the iterator only goes up.
Something like:
```
if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST)
     up_bound = min_t(up_bound, opts.file_validation_chksm_parameter);
if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD)
     step = opts.file_validation_chksm_parameter;
else
     step = 1;
for (i = 0; i < up_bound; i += step)
{
/* ... */
}
```
At this moment the logic is obscure and hard to read.

Reply:
I generally I agree with @0x7f454c46, this code can be simplified. But at this point the logic looks at least semanticly _right_ to me, please don't break it =)

This:
up_bound = min_t(up_bound, opts.file_validation_chksm_parameter);
does not go well with:
up_bound = min_t(size_t, fd_status->st_size, low_bound + CHKSM_CHUNK_SIZE);

So be careful.

Reply:
What I meant was:
```
up_bound = min_t(size_t, fd_status->st_size, low_bound + CHKSM_CHUNK_SIZE);
if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST)
     up_bound = min_t(up_bound, opts.file_validation_chksm_parameter);
```
and some update inside the loop on remap.

Reply:
I've moved some sections of code around, hopefully it's easier to read now. I wrote it this way to make it really easy to add new configurations.
To do so, only the `init, next and stop` functions need to be altered without worrying about the rest of code (Like mapping the next region of the file).
Even if the code for a new configuration is faulty, it won't crash since the iterator is validated outside those functions before accessing memory.